### PR TITLE
Fix #160

### LINKED ISSUE #160: PubSubEngine interface's publish function should not be synchronous
[The PubSubEngine interface](https://github.com/apollographql/graphql-subscriptions/blob/master/src/pubsub-engine.ts#L2) specifies the publish function as:
```
publish(triggerName: string, payload: any): boolean;
```
which is synchronous.

In my opinion, this is inappropriate for any production (i.e., anything backed by an external store) deployment.  In those implementations, publishing often goes over the network, and might fail or otherwise be lost.

The current semantics of `PubSubEngine` aren't well documented, but assuming that the return value of `publish` should indicate a success or failure, they demand that implementations can synchronously determine success or failure, which isn't feasible if they go over the network.

Looking over what the current linked implementations do:
- The Redis implementation has a [TODO saying to file an issue about this](https://github.com/davidyaha/graphql-redis-subscriptions/blob/master/src/redis-pubsub.ts#L62) cc @davidyaha  
- The Google Pubsub implementation [ignores the interface typing](https://github.com/axelspringer/graphql-google-pubsub/blame/master/src/google-pubsub.js#L23) and [returns a promise](https://cloud.google.com/nodejs/docs/reference/pubsub/0.19.x/Publisher#publish) cc @jonas-arkulpa 
- The MQTT implementation creates a [floating promise](https://github.com/davidyaha/graphql-mqtt-subscriptions/blob/master/src/mqtt-pubsub.ts#L51) and [returns true unconditionally](https://github.com/davidyaha/graphql-mqtt-subscriptions/blob/master/src/mqtt-pubsub.ts#L52)
- The RabbitMQ implementation does the same thing: [floats a](https://github.com/cdmbase/graphql-rabbitmq-subscriptions/blob/master/src/amqp-pubsub.ts#L51) [promise](https://github.com/cdmbase/rabbitmq-pubsub/blob/master/src/publisher.ts#L14), [returns true](https://github.com/cdmbase/graphql-rabbitmq-subscriptions/blob/master/src/amqp-pubsub.ts#L52)
- It's a bit hard to disentangle what the [Kafka implementation does](https://github.com/ancashoria/graphql-kafka-subscriptions/blob/master/src/kafka-pubsub.ts#L50), but near as I can tell, it writes to a [NodeJS stream](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback), not supplying an data flush callback.  I think this is morally equivalent to floating a promise, but with the boolean value indicating when the stream isn't draining.
- The [Postgres implementation](https://github.com/GraphQLCollege/graphql-postgres-subscriptions/blob/master/postgres-pubsub.js#L20) uses an event emitter, which I don't think has any guaranteed delivery semantics, and returns true.
- The [NATS implementation](https://github.com/moonwalker/graphql-nats-subscriptions/blob/master/src/nats-pubsub.ts#L17) does not supply a callback to the [publish function](https://github.com/nats-io/node-nats/blob/master/lib/nats.js#L1164) and returns true.

tl:dr; Please make `publish` return a `Promise<boolean>`.

### PR DESCRIPTION
**Pull request by grantwwu**
See discussion in #160

### CODE REVIEW

File: src/pubsub-engine.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 export interface PubSubEngine {
-  publish(triggerName: string, payload: any): boolean;
+  publish(triggerName: string, payload: any): boolean | Promise<void>;
```
Comment (NeoPhi) [inline src/pubsub-engine.ts]: I think we should drop the `boolean` since adding the `Promise` already introduces a breaking change.

Comment (grantwwu) [reply src/pubsub-engine.ts]: 👍

**Comment by NeoPhi** [comment]
@grantwwu Can you address the failing tests?

**Comment by grantwwu** [comment]
Ah, yes, I'll get to that soon.

**Comment by grantwwu** [comment]
Just got this green

File: src/test/tests.ts [hunk 2]
```diff
@@ -26,33 +26,29 @@ describe('PubSub', function() {
...
-    }).then((subId) => {
-      ps.unsubscribe(subId);
-      const succeed = ps.publish('a', 'test');
-      expect(succeed).to.be.true; // True because publish success is not
-                                  // indicated by trigger having subscriptions
-      done(); // works because pubsub is synchronous
-    });
+    let subId = await ps.subscribe('a', payload => { assert(false); });
+    ps.unsubscribe(subId);
+    // console.log("werwwerwerwer");
```
Comment (NeoPhi) [inline src/test/tests.ts]: Unneeded.

File: src/pubsub-engine.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 export interface PubSubEngine {
-  publish(triggerName: string, payload: any): boolean;
+  publish(triggerName: string, payload: any): Promise<void>;
```
Comment (NeoPhi) [inline src/pubsub-engine.ts]: We should document this change in CHANGELOG

Comment (grantwwu) [reply src/pubsub-engine.ts]: Got it, I assume we should bump to 0.6.0?

File: src/test/tests.ts (hunk 2 above)
Comment (grantwwu) [reply src/test/tests.ts]: 😓

File: src/pubsub-engine.ts (hunk 3 above)
Comment (grantwwu) [reply src/pubsub-engine.ts]: I put this under vNext, right?

Comment (NeoPhi) [reply src/pubsub-engine.ts]: Yup.

**Pull request merged**